decentralize-authz-app/src/repositories/kv-repository.ts

```diff
@@ -0,0 +1,186 @@
+  }
+  
+  public get(key: string): ServiceResult<T> {
+    try {
+      const value: any = this.kvStore.get(ccf.strToBuf(key));
+     
+      if (value === undefined) {
+        return ServiceResult.Failed({
+          errorMessage: "Error: key does not exist",
+          errorType: "KeyValueStoreError",
+        });
+      }
+
+      const data = ccf.bufToJsonCompatible(value) as T;
+
+      return ServiceResult.Succeeded(data);
+    } catch (ex) {
+      return ServiceResult.Failed({
+        errorMessage: "Error: unable to read value from the kvstore",
+        errorType: "KeyValueStoreError",
+        details: ex,
+      });
+    }
+  }
+
+  public has(key: string): ServiceResult<boolean> {
+    try {
+      return ServiceResult.Succeeded(this.kvStore.has(ccf.strToBuf(key)));
+    } catch (ex) {
+      return ServiceResult.Failed({
+        errorMessage: "Error: unable to check if key exists in the kvstore",
+        errorType: "KeyValueStoreError",
+        details: ex,
+      });
+    }
+  }
+
+  public keys(): ServiceResult<string[]> {
+    try {
+      const keys: string[] = [];
+      this.kvStore.forEach((val, key) => {
+        keys.push(ccfapp.bufToStr(key));
+      });
+      return ServiceResult.Succeeded(keys);
+    } catch (ex) {
+      return ServiceResult.Failed({
+        errorMessage: "Error: unable to get kvstore all keys",
+        errorType: "KeyValueStoreError",
+        details: ex,
+      });
+    }
+  }
+
+  public values(): ServiceResult<T[]> {
+    try {
+      const values: T[] = [];
+      this.kvStore.forEach((val, key) => {
+        values.push(val);
+      });
+      return ServiceResult.Succeeded(values);
+    } catch (ex) {
+      return ServiceResult.Failed({
+        errorMessage: "Error: unable to get kvstore all values",
+        errorType: "KeyValueStoreError",
+        details: ex,
+      });
+    }
+  }
+
+  public clear(): ServiceResult<void> {
+    try {
+      return ServiceResult.Succeeded(this.kvStore.clear());
+    } catch (ex) {
+      return ServiceResult.Failed({
+        errorMessage: "Error: unable to clear kvstore values",
+        errorType: "KeyValueStoreError",
+        details: ex,
+      });
+    }
+  }
+
+  public delete(key: string): ServiceResult<void> {
+    try {
+      return ServiceResult.Succeeded(this.kvStore.delete(ccf.strToBuf(key)));
+    } catch (ex) {
+      return ServiceResult.Failed({
+        errorMessage: "Error: unable to remove a key",
+        errorType: "KeyValueStoreError",
+        details: ex,
+      });
+    }
+  }
+
+  // 
+  public forEach(callback: (key: string, value: T) => void): ServiceResult<string> {
+    try {
+      this.kvStore.forEach((val, key) => {
+        callback(ccf.bufToStr(key), val);
+      });
+
+      return ServiceResult.Succeeded("");
+    } catch (ex) {
+      return ServiceResult.Failed({
+        errorMessage: "Error: unable to iterate the kvstore pairs",
+        errorType: "KeyValueStoreError",
+        details: ex,
```

Comment:
This does not seem ideal, because there's a good chance the exception if there is one come from the user callback. The error type "KeyValyeStoreError" may be quite misleading here. In addition, if the user exception is verbose, contains confidential information, and "return_exception_details" is set to false in the governance, this will still leak it.

Reply:
Removed the ex and changed the error message